core/src/types/transaction/mod.rs

```diff
@@ -39,6 +40,18 @@ pub fn hash_tx(tx_bytes: &[u8]) -> Hash {
     Hash(*digest.as_ref())
 }
 
+/// Get the hash of the unsigned transaction (if signed), otherwise the hash of
+/// entire tx.
+pub fn unsigned_hash_tx(tx_bytes: &[u8]) -> Hash {
+    match SignedTxData::try_from_slice(tx_bytes) {
```

Comment:
I don't know if this quite right - I think for both wrappers and inner txs, the `tx_bytes` is encoded `Tx`, which has the `SignedTxData` in the `data` field. To hash unsigned tx, you can try to decode `SignedTxData` from the `tx.data` and if it's valid, replace the `tx.data` with the `SignedTxData.data` (this is the unsigned data) field to reconstruct unsigned tx.